src/interpreter.rs

```diff
@@ -526,46 +527,34 @@ impl<'a, 'b, C: ContextObject> Interpreter<'a, 'b, C> {
-                    (true, true)
-                };
-
-                if external {
-                    if let Some((_function_name, function)) = self.executable.get_loader().get_function_registry(self.executable.get_sbpf_version()).lookup_by_key(insn.imm as u32) {
-                        resolved = true;
-
-                        self.vm.due_insn_count = self.vm.previous_instruction_meter - self.vm.due_insn_count;
-                        self.vm.registers[0..6].copy_from_slice(&self.reg[0..6]);
-                        self.vm.invoke_function(function);
-                        self.vm.due_insn_count = 0;
-                        self.reg[0] = match &self.vm.program_result {
-                            ProgramResult::Ok(value) => *value,
-                            ProgramResult::Err(_err) => return false,
-                        };
-                    }
-                }
-
-                if internal && !resolved {
-                    if let Some((_function_name, target_pc)) = self.executable.get_function_registry().lookup_by_key(insn.imm as u32) {
-                        resolved = true;
-
-                        // make BPF to BPF call
-                        if !self.push_frame(config) {
-                            return false;
-                        }
-                        check_pc!(self, next_pc, target_pc as u64);
+            ebpf::CALL_IMM => {
+                if let Some((_, function)) = self.executable.get_loader().get_function_registry(self.executable.get_sbpf_version()).lookup_by_key(insn.imm as u32) {
+                    // SBPFv1 syscall
+                    self.reg[0] = match self.dispatch_syscall(function) {
+                        ProgramResult::Ok(value) => *value,
+                        ProgramResult::Err(_err) => return false,
+                    };
+                } else if let Some((_, target_pc)) = self.executable.get_function_registry().lookup_by_key(insn.imm as u32) {
+                    // make BPF to BPF call
+                    if !self.push_frame(config) {
+                        return false;
                     }
-                }
-
-                if !resolved {
+                    check_pc!(self, next_pc, target_pc as u64);
+                } else {
                     throw_error!(self, EbpfError::UnsupportedInstruction);
                 }
             }
+            ebpf::SYSCALL if self.executable.get_sbpf_version().static_syscalls() => {
+                if let Some((_, function)) = self.executable.get_loader().get_function_registry(self.executable.get_sbpf_version()).lookup_by_key(insn.imm as u32) {
+                    // SBPFv2 syscall
+                    self.reg[0] = match self.dispatch_syscall(function) {
+                        ProgramResult::Ok(value) => *value,
+                        ProgramResult::Err(_err) => return false,
+                    };
+                } else {
+                    throw_error!(self, EbpfError::UnsupportedInstruction);
```

Comment:
This could be a `debug_assert!(false);` because it should be unreachable.

Reply:
True, but I wanted to be conservative and follow the same approach of CALL_IMM.

Either way, I can change here and in the jitter to only an assert.

Reply:
In `test_code_length_estimate`, we check for an `EbpgError::UnsupportedInstruction` when the result is not valid. Adding the debug assert here would break the test. Perhaps, we should keep the error there for consistency. What do you think?

Reply:
(I fixed the test).